k8s/store.go

```diff
@@ -0,0 +1,156 @@
+	GetOrCreateNamespaceSource() (*WatchedResource, error)
+	// GetOrCreateIngressSource create an ingress informer and returns its corresponding store and event handler
+	GetOrCreateIngressSource() (*WatchedResource, error)
+	// GetOrCreateServiceSource create a service informer and returns its corresponding store and event handler
+	GetOrCreateServiceSource() (*WatchedResource, error)
+}
+
+type lazyLoadedStore struct {
+	clientset                *kubernetes.Clientset
+	stopCh                   chan struct{}
+	resyncPeriod             time.Duration
+	informerFactory          informerFactory
+	eventHandlerFactory      eventHandlerFactory
+	namespaceWatchedResource *WatchedResource
+	ingressWatchedResource   *WatchedResource
+	serviceWatchedResource   *WatchedResource
+}
+
+// WatchedResource exposes a store for the resource to allow interrogating the current state via get/list operations
+// and an watcher so clients can subscribe to update notifications
+type WatchedResource struct {
+	store   cache.Store
+	watcher *handlerWatcher
+}
+
+// Implement cache.ResourceEventHandler
+var _ informerFactory = &cacheInformerFactory{}
+
+type informerFactory interface {
+	createNamespaceInformer(time.Duration, cache.ResourceEventHandler) (cache.Store, cache.Controller)
+	createIngressInformer(time.Duration, cache.ResourceEventHandler) (cache.Store, cache.Controller)
+	createServiceInformer(time.Duration, cache.ResourceEventHandler) (cache.Store, cache.Controller)
+}
+
+type cacheInformerFactory struct {
+	clientset *kubernetes.Clientset
+}
+
+// Implement cache.ResourceEventHandler
+var _ informerFactory = &cacheInformerFactory{}
+
+// NewStore creates a new store that starts watching for resources only when requested
+func NewStore(clientset *kubernetes.Clientset, stopCh chan struct{}, resyncPeriod time.Duration) Store {
+	return &lazyLoadedStore{
+		clientset:           clientset,
+		stopCh:              stopCh,
+		resyncPeriod:        resyncPeriod,
+		informerFactory:     &cacheInformerFactory{clientset: clientset},
+		eventHandlerFactory: &bufferedEventHandlerFactory{},
+	}
+}
+
+//  GetOrCreateNamespaceSource creates a namespace informer and registers and event handler to watch for changes
+func (s *lazyLoadedStore) GetOrCreateNamespaceSource() (*WatchedResource, error) {
+	if s.namespaceWatchedResource != nil {
+		return s.namespaceWatchedResource, nil
+	}
+
+	log.Debug("Creating an informer to watch namespace resources")
+	namespaceWatcher := s.eventHandlerFactory.createBufferedHandler(bufferedWatcherDuration)
+	store, controller := s.informerFactory.createNamespaceInformer(s.resyncPeriod, namespaceWatcher)
+
+	go controller.Run(s.stopCh)
+
+	if !cache.WaitForCacheSync(s.stopCh, controller.HasSynced) {
+		return &WatchedResource{}, errors.New("error while waiting for namespace cache to populate")
+	}
+
+	log.Debug("Namespace cache has been fully populated")
+	return &WatchedResource{
+		store:   store,
+		watcher: namespaceWatcher,
+	}, nil
```

Comment:
I think there's scope for refactoring this out into code shared with the subsequent two methods. It all looks common apart from the `informerFactory` call on line 82, but that function could be passed as an argument into an extracted common function.

Reply:
sounds good